### PR TITLE
Cleanup: Move instead of copy and use XMLWriter for DGML

### PR DESCRIPTION
**Pull request by Thomas1664**
- `BinaryParagraph(Paragraph)` --> `BinaryParagraph(Paragraph&&)`
- Add missing &
- Use `std::move`
- Initialize string instead of default-initialize + append

### CODE REVIEW

File: src/vcpkg/commands.cache.cpp [hunk 1]
```diff
@@ -15,11 +15,10 @@ namespace vcpkg::Commands::Cache
         std::vector<BinaryParagraph> output;
         for (auto&& path : paths.get_filesystem().get_files_non_recursive(paths.packages(), VCPKG_LINE_INFO))
         {
-            const auto pghs = Paragraphs::get_single_paragraph(paths.get_filesystem(), path / "CONTROL");
-            if (const auto p = pghs.get())
+            auto pghs = Paragraphs::get_single_paragraph(paths.get_filesystem(), std::move(path) / "CONTROL");
+            if (auto p = pghs.get())
```
Comment (BillyONeal) [inline src/vcpkg/commands.cache.cpp:19]:
```suggestion
            if (const auto p = pghs.get())
```

The pointer can be const.

File: src/vcpkg/commands.check-support.cpp [hunk 2]
```diff
@@ -147,7 +147,7 @@ namespace vcpkg::Commands
             bool user_supported = false;
 
             std::vector<Port> dependencies_not_supported;
-            for (const auto& action : action_plan.install_actions)
+            for (auto&& action : action_plan.install_actions)
```
Comment (BillyONeal) [inline src/vcpkg/commands.check-support.cpp:150]: I'm not sure I agree with this change; the action plan is not obviously 'consumed' here.

Rearranging things such that the action_plan is destroyed after this for loop body would resolve my concern.

File: src/vcpkg/commands.ci-verify-versions.cpp [hunk 3]
```diff
@@ -150,7 +150,7 @@ namespace vcpkg::Commands::CIVerifyVersions
         {
             return {msg::format_error(msgWhileLoadingLocalPort, msg::package_name = port_name)
                         .append_raw('\n')
-                        .append_raw(maybe_scf.error()->error),
+                        .append_raw(std::move(maybe_scf).error()->error),
```
Comment (BillyONeal) [inline src/vcpkg/commands.ci-verify-versions.cpp:153]: This is OK but has no effect.

File: src/vcpkg/paragraphs.cpp [hunk 4]
```diff
@@ -482,11 +482,11 @@ namespace vcpkg::Paragraphs
...
         {
             BinaryControlFile bcf;
-            bcf.core_paragraph = BinaryParagraph(p->front());
+            bcf.core_paragraph = BinaryParagraph(std::move(p->front()));
             p->erase(p->begin());
 
-            bcf.features =
-                Util::fmap(*p, [&](auto&& raw_feature) -> BinaryParagraph { return BinaryParagraph(raw_feature); });
+            bcf.features = Util::fmap(
+                *p, [&](auto&& raw_feature) -> BinaryParagraph { return BinaryParagraph(std::move(raw_feature)); });
```
Comment (BillyONeal) [inline src/vcpkg/paragraphs.cpp:489]: I don't believe this change is correct. Note that `auto&&` is perfect forwarding, not move.

File: src/vcpkg/commands.create.cpp [hunk 5]
```diff
@@ -45,13 +45,13 @@ namespace vcpkg::Commands::Create
...
         {
-            const std::string& zip_file_name = parsed.command_arguments[2];
+            std::string& zip_file_name = parsed.command_arguments[2];
             Checks::msg_check_exit(VCPKG_LINE_INFO,
                                    !has_invalid_chars_for_filesystem(zip_file_name),
                                    msgInvalidFilename,
                                    msg::value = FILESYSTEM_INVALID_CHARACTERS,
                                    msg::path = zip_file_name);
-            cmake_args.emplace_back("FILENAME", zip_file_name);
+            cmake_args.emplace_back("FILENAME", std::move(zip_file_name));
```
Comment (BillyONeal) [inline src/vcpkg/commands.create.cpp:54]: I'm very nervous about destroying command_arguments by proxy like this; I don't think saving this 1 parameter copy is worth that.

File: src/vcpkg/commands.dependinfo.cpp [hunk 6]
```diff
@@ -169,10 +169,9 @@ namespace vcpkg::Commands::DependInfo
 
     std::string create_dot_as_string(const std::vector<PackageDependInfo>& depend_info)
     {
-        int empty_node_count = 0;
+        unsigned int empty_node_count = 0;
```
Comment (BillyONeal) [inline src/vcpkg/commands.dependinfo.cpp:172]: This change seems unmotivated?

File: src/vcpkg/commands.dependinfo.cpp [hunk 7]
```diff
@@ -197,14 +196,13 @@ namespace vcpkg::Commands::DependInfo
 
     std::string create_dgml_as_string(const std::vector<PackageDependInfo>& depend_info)
     {
-        std::string s;
-        s.append("<?xml version=\"1.0\" encoding=\"utf-8\"?>");
-        s.append("<DirectedGraph xmlns=\"http://schemas.microsoft.com/vs/2009/dgml\">");
+        std::string s =
```
Comment (BillyONeal) [inline src/vcpkg/commands.dependinfo.cpp:199]: Separating the declaration was intentional. I think if this is being touched it should be changed to use the real XML writer instead.

**Review by BillyONeal — CHANGES_REQUESTED** [review]
This is mostly OK but I think there are 1 or 2 mistakes.
There are others that I *think* are correct but of which I am not in favor for maintainability reasons; maybe other contributors have different opinions on them.

File: src/vcpkg/paragraphs.cpp (hunk 4 above)
Comment (Thomas1664) [reply src/vcpkg/paragraphs.cpp:489]: Doesn't compile without move

File: src/vcpkg/commands.dependinfo.cpp (hunk 7 above)
Comment (Thomas1664) [reply src/vcpkg/commands.dependinfo.cpp]: Done

File: src/vcpkg/paragraphs.cpp (hunk 4 above)
Comment (BillyONeal) [reply src/vcpkg/paragraphs.cpp:489]: I see. How about something like:

```
        ExpectedL<std::vector<Paragraph>> maybe_paragraphs = get_paragraphs(fs, package_dir / "CONTROL");
        if (auto pparagraphs = maybe_paragraphs.get())
        {
            auto& paragraphs = *pparagraphs;
            BinaryControlFile bcf;
            bcf.core_paragraph = BinaryParagraph(std::move(paragraphs[0]));
            if (bcf.core_paragraph.spec != spec)
            {
                return msg::format(msgMismatchedSpec,
                                   msg::path = package_dir,
                                   msg::expected = spec,
                                   msg::actual = bcf.core_paragraph.spec);
            }

            bcf.features.reserve(paragraphs.size() - 1);
            for (std::size_t idx = 1; idx < paragraphs.size(); ++idx)
            {
                bcf.features.emplace_back(BinaryParagraph{std::move(paragraphs[idx])});
            }

            return bcf;
        }

        return maybe_paragraphs.error();
```
?

Comment (Thomas1664) [reply src/vcpkg/paragraphs.cpp]: My understanding is that `reserve()` specifies the absolute capacity of the vector and not `size() + new_capacity`

[cppreference](https://en.cppreference.com/w/cpp/container/vector/reserve):

> Increase the capacity of the vector (the total number of elements that the vector can hold without requiring reallocation) **to** a value that's greater or equal to new_cap.

Note that it says __to__ as opposed to __by__.

Comment (Thomas1664) [reply src/vcpkg/paragraphs.cpp]: Thanks for the suggestion!

**Review by BillyONeal — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks for your contribution!

**Pull request merged**